lib/mongo/socket.rb

```diff
@@ -334,12 +334,36 @@ def read_from_socket(length, timeout: nil)
             raise Errno::ETIMEDOUT, "Took more than #{_timeout} seconds to receive data"
           end
         end
+        fd = options[:fd]
         if exc.is_a?(IO::WaitReadable)
-          select_args = [[@socket], nil, [@socket], select_timeout]
+          if fd
+            select_args = [[@socket, fd], nil, [@socket, fd], select_timeout]
+          else
+            select_args = [[@socket], nil, [@socket], select_timeout]
+          end
         else
           select_args = [nil, [@socket], [@socket], select_timeout]
         end
+
         rv = Kernel.select(*select_args)
```

Comment:
We should catch and IO error here, shouldn't we? Or is it propagated?

Reply:
it's propagated to the `map_exceptions` method where it's converted to a `Mongo::SocketError`